### PR TITLE
OTA Receiver struct + Metadata for trades

### PR DESCRIPTION
**Pull request by dat-incognito-org**
- define new metadata types + constants for trade & addOrder flow
- define OTAReceiver struct as `privacy` substitute for PaymentAddress when sending coin
- move `calculateSize` to `metadata/common`
- future work : withdrawOrder metadata; validation functions

### CODE REVIEW

File: metadata/pdexv3/addOrder.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	return req.Type == metadataCommon.PDexV3AddOrderRequestMeta
+}
+
+func (req AddOrderRequest) Hash() *common.Hash {
+	rawBytes, _ := json.Marshal(req)
+	hash := common.HashH([]byte(rawBytes))
+	return &hash
+}
+
+func (req *AddOrderRequest) BuildReqActions(tx metadataCommon.Transaction, chainRetriever metadataCommon.ChainRetriever, shardViewRetriever metadataCommon.ShardViewRetriever, beaconViewRetriever metadataCommon.BeaconViewRetriever, shardID byte, shardHeight uint64) ([][]string, error) {
```
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go]: fixed

File: metadata/pdexv3/addOrder.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+// AddOrderRequest
+type AddOrderRequest struct {
+	TokenToSell         common.Hash         `json:"TokenToSell"`
+	TokenToBuy          common.Hash         `json:"TokenToBuy"`
+	SellAmount          uint64              `json:"SellAmount"`
+	MinAcceptableAmount uint64              `json:"MinAcceptableAmount"`
+	TradingFee          uint64              `json:"TradingFee"`
+	RefundReceiver      privacy.OTAReceiver `json:"RefundReceiver"`
```
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go:21]: for addOrder, success means the new order will be stored, no funds involved

File: metadata/pdexv3/addOrder.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	metadataCommon.MetadataBase
+}
+
+type AddOrderAction struct {
```
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go:25]: will remove later

File: metadata/pdexv3/trade.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	TokenToBuy          common.Hash         `json:"TokenToBuy"`
+	SellAmount          uint64              `json:"SellAmount"`
+	MinAcceptableAmount uint64              `json:"MinAcceptableAmount"`
+	TradingFee          uint64              `json:"TradingFee"`
+	Receiver            privacy.OTAReceiver `json:"Receiver"`
+	RefundReceiver      privacy.OTAReceiver `json:"RefundReceiver"`
+	metadataCommon.MetadataBase
+}
+
+type TradeAction struct {
```
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/trade.go:27]: will remove later

File: metadata/pdexv3/addOrder.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	ShardID     byte        `json:"ShardID"`
+	RequestTxID common.Hash `json:"RequestTxID"`
+}
+
+type RefundedAddOrder struct {
+	Receiver    privacy.OTAReceiver `json:"Receiver"`
+	TokenToSell common.Hash         `json:"TokenToSell"`
+	Amount      uint64              `json:"Amount"`
+	ShardID     byte                `json:"ShardID`
+	RequestTxID common.Hash         `json:"RequestTxID"`
```
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go:43]: what are we trying to match this particular name with ?
also, if the name turns out to be relevant, it should be explicitly stated using a getter interface

File: metadata/pdexv3/addOrder.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+func NewAddOrderRequest(
+	tokenToBuy common.Hash,
+	tokenToSell common.Hash,
+	sellAmount uint64,
+	minAcceptableAmount uint64,
+	tradingFee uint64,
+	refundRecv privacy.OTAReceiver,
+	metaType int,
+) (*AddOrderRequest, error) {
```
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go:54]: good question. Might be a reserved error slot; many other metadata types have this constructor prototype, so i don't intend to change it right now

File: metadata/pdexv3/addOrder.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tin-incognito) [reply metadata/pdexv3/addOrder.go:43]: ok we will do that later

File: metadata/pdexv3/addOrder.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tin-incognito) [reply metadata/pdexv3/addOrder.go:21]: So the user will receive tokenWantToBuy later by `RefundReceiver` or other mechanism?

File: metadata/pdexv3/addOrder.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	ShardID     byte            `json:"ShardID"`
+	RequestTxID common.Hash     `json:"RequestTxID"`
+}
+
+type AcceptedAddOrder struct {
+	TokenToBuy  common.Hash `json:"TokenToBuy"`
+	PairID      string      `json:"PairID"`
+	ShardID     byte        `json:"ShardID"`
+	RequestTxID common.Hash `json:"RequestTxID"`
+}
```
Comment (cuong-incognito) [inline metadata/pdexv3/addOrder.go:36]: Thought there should be more fields such as `Amount`, `Price`

File: metadata/pdexv3/trade.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+// TradeRequest
+type TradeRequest struct {
+	TokenToSell         common.Hash         `json:"TokenToSell"`
+	TokenToBuy          common.Hash         `json:"TokenToBuy"`
+	SellAmount          uint64              `json:"SellAmount"`
+	MinAcceptableAmount uint64              `json:"MinAcceptableAmount"`
+	TradingFee          uint64              `json:"TradingFee"`
+	Receiver            privacy.OTAReceiver `json:"Receiver"`
+	RefundReceiver      privacy.OTAReceiver `json:"RefundReceiver"`
+	metadataCommon.MetadataBase
```
Comment (cuong-incognito) [inline metadata/pdexv3/trade.go:24]: We don't accept trade with a router anymore, do we?

File: metadata/pdexv3/addOrder.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go:22]: - the user creates a `WithdrawOrderRequest`, which will contain the `Receiver` for their order's _bought_ token
- see more in the updated notes in document
- in this PR the constant for WithdrawOrderRequestMeta has been declared, but the type itself not yet

File: metadata/pdexv3/addOrder.go (hunk 7 above)
Comment (dat-incognito-org) [reply metadata/pdexv3/addOrder.go]: added

File: metadata/pdexv3/trade.go (hunk 8 above)
Comment (dat-incognito-org) [reply metadata/pdexv3/trade.go]: good catch
changed both token fields to `TradePath`

**Review by cuong-incognito — APPROVED** [review]

**Review by tin-incognito — APPROVED** [review]
LGTM!

**Pull request merged**